src/core/trulens/experimental/otel_tracing/core/instrument.py

```diff
@@ -0,0 +1,83 @@
+from functools import wraps
+import logging
+from typing import Any, Callable, Optional, Union
+
+from opentelemetry import trace
+from trulens.experimental.otel_tracing.core.init import TRULENS_SERVICE_NAME
+from trulens.experimental.otel_tracing.core.semantic import (
+    TRULENS_SELECTOR_NAME,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def instrument(
+    attributes: Optional[
+        Union[dict[str, Any], Callable[[Any, Any, Any], dict[str, Any]]]
+    ] = {},
+):
+    """
+    Decorator for marking functions to be instrumented in custom classes that are
+    wrapped by TruCustomApp, with OpenTelemetry tracing.
+    """
+
+    def _validate_selector_name(final_attributes: dict[str, Any]):
+        if TRULENS_SELECTOR_NAME in final_attributes:
+            selector_name = final_attributes[TRULENS_SELECTOR_NAME]
+            if not isinstance(selector_name, str):
+                raise ValueError(
+                    f"Selector name must be a string, not {type(selector_name)}"
+                )
+
+    def _validate_attributes(final_attributes: dict[str, Any]):
+        _validate_selector_name(final_attributes)
+        # TODO: validate OTEL attributes.
+        # TODO: validate span type attributes.
+
+    def inner_decorator(func: Callable):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            with (
+                trace.get_tracer_provider()
+                .get_tracer(TRULENS_SERVICE_NAME)
+                .start_as_current_span(
+                    name=func.__name__,
+                )
+            ) as parent_span:
+                span = trace.get_current_span()
+
+                span.set_attribute("name", func.__name__)
+                span.set_attribute("kind", "SPAN_KIND_TRULENS")
+                span.set_attribute(
+                    "parent_span_id", parent_span.get_span_context().span_id
+                )
+
+                ret = func(*args, **kwargs)
+
+                try:
+                    attributes_to_add = {}
+
+                    # Set the user provider attributes.
+                    if attributes:
+                        if callable(attributes):
+                            attributes_to_add = attributes(ret, *args, **kwargs)
+                        else:
+                            attributes_to_add = attributes
+
+                    logger.info(f"Attributes to add: {attributes_to_add}")
+
+                    _validate_attributes(attributes_to_add)
+
+                    for key, value in attributes_to_add.items():
+                        span.set_attribute(key, value)
```

Comment:
this won't prepend any namespaces but maybe it's just okay to get the user to say the full attribute name (i.e. with the scopes) themselves. If that's your idea, I'm good with that (so long as later when we validate spans with a given span type follow the semantic convention of that span type and throw an error I'm good).

Reply:
update to prepend namespace